### PR TITLE
Add ADR 0004 - Naming of Interface Implementations

### LINKED ISSUE #7718: Decide on prefixes for app-specific classes/function names
We need to decide on prefixes for app-specific classes for both K-9 Mail and Thunderbird for Android. This is mostly for app-specific implementations of an interface. But maybe also for app-specific compose functions, e.g. `K9Theme` and `TbTheme` (although in this case maybe having them called `K9MailTheme` and `ThunderbirdTheme` is not a big deal, because those are usually used on a single line together with only an opening brace).

I'm suggesting `K9` for K-9 Mail and `Tb` for Thunderbird for Android. These have the advantage of being short. Which feels important when the interface name already has 3+ components, e.g. `OAuthConfigurationFactory` (→ `K9OAuthConfigurationFactory`/`TbOAuthConfigurationFactory` vs. `K9MailOAuthConfigurationFactory`/`ThunderbirdOAuthConfigurationFactory`).

### LINKED ISSUE #7719: Decide on generic prefix for the default implementation of an interface
Right now we have lots of interfaces that have only one implementation. The name of the implementation for the interface `Interface` is usually `PrefixInterface`. But we use quite a few different values for the prefix (`K9`, `Real`, `Default`). 

Whenever possible we should use a descriptive prefix, e.g. `DatabaseMessageStore` for a `MessageStore` implementation that stores messages in a database.

However, sometimes it's not easy to come up with a descriptive prefix. In those cases we should use a standard prefix. And it probably shouldn't be `K9`, unless the implementation is specific to K-9 Mail (see #7718).

Once we've decided on a prefix, we should rename existing classes that use a different generic prefix.

### PR DESCRIPTION
**Pull request by cketti**
Closes #7718
Closes #7719

### CODE REVIEW

File: docs/architecture/adr/0004-naming-of-interface-implementations.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Naming of Interface Implementations
```
Comment (wmontwe) [inline docs/architecture/adr/0004-naming-of-interface-implementations.md:1]: Alternative

```suggestion
# Naming Conventions for Interfaces and Their Implementations
```

File: docs/architecture/adr/0004-naming-of-interface-implementations.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+interface and the implementations (e.g. the interface `Backend` with the implementations `ImapBackend` and
+`Pop3Backend`). Naming becomes harder when the interface mainly exists to allow having isolated unit tests and the
+production code contains exactly one implementation of the interface.
+Prior to this ADR we didn't have any naming guidelines and the names varied widely. Often when there was only one
+(production) implementation, the class name used one of the prefixes `Default`, `Real`, or `K9`. None of these had any
+special meaning and it wasn't clear which one to pick when creating a new interface/class pair.
+
+## Decision
+
+We'll be using the following guidelines for naming interfaces and classes implementing an interface:
```
Comment (wmontwe) [inline docs/architecture/adr/0004-naming-of-interface-implementations.md:19]:
```suggestion
We'll be using the following guidelines for naming interfaces and their implementation classes:
```

File: docs/architecture/adr/0004-naming-of-interface-implementations.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+- Name interfaces like you would name a class. Do not use the `IInterface` pattern. Try to keep the name short.
+- Name implementations by adding a prefix to the interface name, e.g. `InMemoryThing` for an interface named `Thing`.
+- Try to use a descriptive name whenever possible, e.g. `DatabaseMessageStore` instead of `DefaultMessageStore`.
+- When adding an interface to isolate platform specific behavior, use the platform name as the prefix, e.g.
+  `AndroidPowerManager`.
+- When there are app-specific implementations of an interface, use the prefix `K9` for K-9 Mail and `Tb` for
+  Thunderbird. Example: `K9AppNameProvider` and `TbAppNameProvider`.
+- When there's no (short) descriptive name and there's only one production implementation of an interface, use the
+  prefix `Default`. Example: `DefaultImapFolder`.
```
Comment (wmontwe) [inline docs/architecture/adr/0004-naming-of-interface-implementations.md:29]: I added highlight for the sections to give a better overview and changed the descriptions.

```suggestion
1. **Interface Naming:** Name interfaces as if they were classes, using a clear and descriptive name. Avoid using the "IInterface" pattern.
2. **Implementation Naming:* Use a prefix that clearly indicates the relationship between the interface and implementation, such as `DatabaseMessageStore` or `InMemoryMessageStore` for the `MessageStore` interface.
3. **Descriptive Names:** Use descriptive names for interfaces and implementing classes that accurately reflect their purpose and functionality.
4. **Platform-specific Implementations:** Use the platform name as a prefix for interface implementations specific to that platform, e.g. `AndroidPowerManager`.
5. **App-specific Implementations:** Use the prefix`K9` for K-9 Mail and `Tb` for
  Thunderbird when app-specific implementations are needed, e.g. `K9AppNameProvider` and `TbAppNameProvider`.
6. **Flexibility:** If no brief descriptive name fits and there is only one production implementation, use the prefix `Default`, like `DefaultImapFolder`.
```

File: docs/architecture/adr/0004-naming-of-interface-implementations.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+## Consequences
+
+- **Positive Consequences**
+
+  - This gives us more consistent naming of interface implementations.
```
Comment (wmontwe) [inline docs/architecture/adr/0004-naming-of-interface-implementations.md:35]:
```suggestion
- Improved code readability and maintainability through consistent naming.
- Reduced confusion and misunderstandings by using clear and descriptive names.
```

File: docs/architecture/adr/0004-naming-of-interface-implementations.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+- **Negative Consequences**
+
+  - Some initial effort is required to rename existing classes that don't follow this naming scheme.
```
Comment (wmontwe) [inline docs/architecture/adr/0004-naming-of-interface-implementations.md:39]:
```suggestion
  - Initial effort is required to rename existing classes that do not follow these naming conventions.
```

**Comment by cketti** [comment]
I applied the suggested changes and reformatted the result to keep Spotless happy.

**Review by wmontwe — APPROVED** [review]
👍

**Pull request merged**